File: cloud-collect-core/src/main/java/com/windea/demo/cloudcollect/core/repository/UserRepository.java
package com.windea.demo.cloudcollect.core.repository;

import com.windea.demo.cloudcollect.core.domain.entity.User;
import com.windea.demo.cloudcollect.core.domain.enums.Role;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;

import java.util.Optional;

public interface UserRepository extends JpaRepository<User, Long> {
	Optional<User> findByUsername(String username);

	Optional<User> findByEmail(String email);

	Page<User> findByNicknameContains(String nickname, Pageable pageable);

	Page<User> findByRole(Role role, Pageable pageable);

	@Query("from User u, in (u.followToUserList) fu where fu.id=:followToUserId")
	Page<User> findByFollowToUser_Id(Long followToUserId, Pageable pageable);

	@Query("select count(u) from User u, in (u.followToUserList) fu where fu.id=:followToUserId")
	long countByFollowToUser_Id(Long followToUserId);

	@Query("from User u, in (u.followByUserList) fu where fu.id=:followByUserId")
	Page<User> findByFollowByUser_Id(Long followByUserId, Pageable pageable);

	@Query("select count(u) from User u, in (u.followByUserList) fu where fu.id=:followByUserId")
	long countByFollowByUser_Id(Long followByUserId);

	@Query("from User u, in(u.praiseToCollectList) c where c.id=:praiseToCollectId")
	Page<User> findByPraiseToCollect_Id(Long praiseToCollectId, Pageable pageable);

	@Query("select count(u) from User u, in(u.praiseToCollectList) c where c.id=:praiseToCollectId")
	long countByPraiseToCollect_Id(Long praiseToCollectId);

	boolean existsByUsernameOrEmail(String username, String email);
}

File: cloud-collect-core/src/main/kotlin/com/windea/demo/cloudcollect/core/repository/UserRepository.kt
package com.windea.demo.cloudcollect.core.repository

import com.windea.demo.cloudcollect.core.domain.entity.*
import com.windea.demo.cloudcollect.core.enums.*
import org.springframework.data.domain.*
import org.springframework.data.jpa.repository.*

interface UserRepository : JpaRepository<User, Long> {
	fun findByUsername(username: String): User?
	
	fun findByEmail(email: String): User?
	
	fun findAllByNicknameContains(nickname: String, pageable: Pageable): Page<User>
	
	fun findAllByUsernameContains(username: String, pageable: Pageable): Page<User>
	
	fun findAllByEmailContains(email: String, pageable: Pageable): Page<User>
	
	fun findAllByRole(role: Role, pageable: Pageable): Page<User>
	
	fun findAllByFollowToUsersId(followToUserId: Long, pageable: Pageable): Page<User>
	
	fun findAllByNicknameContainsAndFollowToUsersId(nickname: String, followToUserId: Long, pageable: Pageable): Page<User>
	
	fun findAllByFollowByUsersId(followByUserId: Long, pageable: Pageable): Page<User>
	
	fun findAllByNicknameContainsAndFollowByUsersId(nickname: String, followByUserId: Long, pageable: Pageable): Page<User>
	
	fun findAllByPraiseToCollectsId(praiseToCollectId: Long, pageable: Pageable): Page<User>
	
	fun countByFollowByUsersId(followByUserId: Long): Long
	
	fun countByFollowToUsersId(followToUserId: Long): Long
	
	fun countByPraiseToCollectsId(praiseToCollectId: Long): Long
	
	fun existsByUsername(username: String): Boolean
	
	fun existsByEmail(email: String): Boolean
	
	fun existsByUsernameOrEmail(username: String, email: String): Boolean
	
	fun existsByIdAndFollowByUsers(id: Long, followByUser: User): Boolean
}
